### DISCUSSION TITLE
Beta 2 no longer waits for stylesheets to load before switching route component?

### DISCUSSION THREAD

**Discussion by yume-chan** [solidjs/solid-start#1403] [Q&A]
I'm migrating my old Solid Start project to beta 2, and I noticed now it has a huge FOUC when switching routes.

I checked the old code, seems like it inserts all `<link rel="stylesheet" >` into `<head>`, wait for them to load, then replaces route content.

Why doesn't it do that now? I think the user experience is worse especially with slow networks (I know it adds preloads, but the user can still switch routes before preloads are loaded)

**Comment by yume-chan** [comment]
This feature was provided by Vite: https://github.com/vitejs/vite/blob/c0ec6bea69b6160553f4a5b30652dcef891788fc/packages/vite/src/node/plugins/importAnalysisBuild.ts

What benefits does nitro/vinxi's manual asset handling provide?

**Comment by ryansolid** [comment]
I don't think Vite waits.. the router and vite have no way of knowing about each other. Vite does handle adding it but if it was loading beforehand it was by chance I think.

**Reply by yume-chan** [reply]
That plugin works on all dynamic imports, old Solid Start file system router generates dynamic imports, so it works out-of-the-box.

---

In fact, in current Solid Start version, you only need to replace

https://github.com/solidjs/solid-start/blob/480802f7d7fdd8aa7f1e018e1433f707ef99601d/packages/start/src/router/lazyRoute.ts#L40

with 

```ts
const mod = await component.build()
```

Vinxi generates the `build` field with Vite-compatible dynamic imports.

---

I created this minimal repro: https://github.com/yume-chan/solid-start-wait-assets

If you start it normally with `pnpm i && pnpm build && pnpm start`, turn on disable cache and network throttling in Chrome DevTools, when it automatically navigates to `/about`, you will see the `<h1>` changes color.

https://github.com/solidjs/solid-start/assets/1330321/f77fea04-0e47-4987-be46-6883f12b648e

But if you execute `node ./patch-solid-start.js` then `pnpm build && pnpm start` again, the navigation will wait for styles to be loaded before rendering the elements. (the script also replaces the dev mode code, which it shouldn't, but doesn't matter in this case)

https://github.com/solidjs/solid-start/assets/1330321/3b02d818-ae56-46be-9da9-38518ac0c5b3

**Reply by yume-chan** [reply]
With minification disabled, the generated router manifest is

```js
const __vite__fileDeps=["assets/about-BzjT0PMC.js","assets/index-CXvkaZX7.js","assets/about-CyiT_ioQ.css","assets/index-CuOQb2Au.js","assets/routing-CH2plF5P.js","assets/index-7whypOeP.css","assets/_...404_-DQ7qBchp.js","assets/HttpStatusCode-B9tIHqYd.js"],__vite__mapDeps=i=>i.map(i=>__vite__fileDeps[i]);

// [omitted]

const fileRoutes = [{
  "$component": {
    "src": "src/routes/about.tsx?pick=default&pick=$css",
    "build": () => __vitePreload(() => import(
      /* @vite-ignore */
      './about-BzjT0PMC.js'
    ),true?__vite__mapDeps([0,1,2]):void 0),
    "import": () => {
      const id = "src/routes/about.tsx?pick=default&pick=$css";
      return __vitePreload(() => import(
        /* @vite-ignore */
        globalThis.MANIFEST["client"].inputs[id].output.path
      ),true?[]:void 0);
    }
  },
  "path": "/about",
  "filePath": "D:/dev/sandbox/wait-assets/wait-assets/src/routes/about.tsx"

// [omitted]

}];
```

in `$component.build`, `__vitePreload` will load and wait `__vite__mapDeps([0,1,2])` (which expands to `['assets/about-BzjT0PMC.js', 'assets/index-CXvkaZX7.js', 'assets/about-CyiT_ioQ.css' ]`). But in `$component.import`, the deps array is empty.

---

Source code of `__vitePreload` is at https://github.com/vitejs/vite/blob/c0ec6bea69b6160553f4a5b30652dcef891788fc/packages/vite/src/node/plugins/importAnalysisBuild.ts#L74-L153
